lib/merkle-tree.js

```diff
@@ -219,40 +213,30 @@ class MerkleTreeBatch {
   commit () {
     if (!this.commitable()) throw INVALID_OPERATION('Tree was modified during batch, refusing to commit')
 
-    if (this.upgraded) this._commitUpgrade()
+    const dbBatch = this.tree.storage.createWriteBatch()
+
+    if (this.upgraded) this._commitUpgrade(dbBatch)
 
     for (let i = 0; i < this.nodes.length; i++) {
       const node = this.nodes[i]
-      this.tree.unflushed.set(node.index, node)
+      dbBatch.addTreeNode(this.batchId, node)
     }
+
+    return dbBatch.flush()
   }
 
-  _commitUpgrade () {
+  _commitUpgrade (dbBatch) {
     // TODO: If easy to detect, we should refuse an trunc+append here without a fork id
     // change. Will only happen on user error so mostly to prevent that.
 
     if (this.ancestors < this.treeLength) {
-      if (this.ancestors > 0) {
-        const head = 2 * this.ancestors
-        const ite = flat.iterator(head - 2)
-
-        while (true) {
-          if (ite.contains(head) && ite.index < head) {
-            this.tree.unflushed.set(ite.index, blankNode(ite.index))
-          }
-          if (ite.offset === 0) break
-          ite.parent()
-        }
-      }
+      dbBatch.deleteTreeNodeRange(this.batchId, this.ancestors, this.treeLengthq)
```

Comment:
you also need to queue the nodes that are gaps now. thats what the code on the left here does, but instead of setting blank nodes you can just queue a deletion obvs

ie
   
```
     3
 1      5
0 2  4 6
```

and you truncate 6 out, then you need to remove 3 also